Harmonize Keycloak Client Configurations with Built-in IdP

michaelstingl opened the pull request:
## Description
This PR harmonizes the Keycloak client configurations with the built-in IdP client configurations to ensure a consistent authentication experience regardless of which identity provider is used.

## Related Issue
- Fixes #588

## Motivation and Context
When switching between the built-in IdP and Keycloak, users may experience authentication issues due to inconsistent client configurations. This PR ensures that client IDs, redirect URIs, domain names, and authentication methods are aligned between both systems.

Key changes:
1. **Standardized client IDs** to match built-in IdP:
   - `web` (unchanged)
   - `OpenCloudDesktop` (was random ID)
   - `OpenCloudAndroid` (was random ID)
   - `OpenCloudIOS` (was random ID)
   - `Cyberduck` (simplified ID)

2. **Fixed domain names**:
   - Changed mobile app URLs from opencloud.com to opencloud.eu
   - Replaced hardcoded domains with template variables

3. **Standardized redirect URIs**:
   - Added specific callback paths for web client
   - Used consistent formats for desktop client

4. **Aligned authentication methods**:
   - Set mobile and desktop clients as public clients (equivalent to native app type)

## How Has This Been Tested?
- Test environment: Local Docker Compose setup with Keycloak enabled
- Test case 1: Login with web client using templated URLs
- Test case 2: Verified client configuration in Keycloak admin console
- Test case 3: Tested template variable replacement in docker-entrypoint-override.sh

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation added

**Note on documentation**: The Keycloak documentation file at `/docs/docs/admin/30-configuration/30-keycloak.md` exists in the separate docs repository as a draft. After this PR is merged, a separate PR should be created in the docs repository to update the documentation with the new client configuration details.

## Detailed Analysis

### 1. Client ID Differences

| Client Type | Built-in IdP ID | Keycloak ID |
|-------------|----------------|-------------|
| Web | `web` | `web` ✅ |
| Desktop | `OpenCloudDesktop` | `xdXOt13JKxym1B1QcEncf2XDkLAexMBFwiT9j6EfhhHFJhs2KM9jbjTmf8JBXE69` ❌ |
| Android | `OpenCloudAndroid` | `e4rAsNUSIUs0lF4nbv9FmCeUkTlV9GdgTLDH1b5uie7syb90SzEVrbN7HIpmWJeD` ❌ |
| iOS | `OpenCloudIOS` | `mxd5OQDk6es5LzOzRvidJNfXLUZS2oN3oUFeXPP8LpPrhx3UroJFduGEYIBOxkY1` ❌ |

### 2. Domain Name Inconsistencies

| Client Type | Built-in IdP Domain | Keycloak Domain |
|-------------|-------------------|----------------|
| Web | `{{OC_URL}}` (templated) | `https://cloud.opencloud.test` (hardcoded) ❌ |
| Android | `oc://android.opencloud.eu` | `oc://android.opencloud.com` ❌ |
| iOS | `oc://ios.opencloud.eu` | `oc://ios.opencloud.com` ❌ |

### 3. Redirect URI Format Differences

| Client Type | Built-in IdP Format | Keycloak Format |
|-------------|-------------------|----------------|
| Web | Template with specific paths | Single wildcard path ❌ |
| Desktop | Fixed ports | Wildcard ports ❌ |

### 4. Client Authentication Type

| Client Type | Built-in IdP | Keycloak |
|-------------|-------------|---------|
| Web | `Trusted: true` | `publicClient: true` (equivalent) ✅ |
| Desktop | `ApplicationType: native` | Uses client secret ❌ |
| Android | `ApplicationType: native` | Uses client secret ❌ |
| iOS | `ApplicationType: native` | Uses client secret ❌ |

micbar commented:
@michaelstingl Thanks! I tested your changes - the clients are working fine! 👏 

One issue is, that the deployment only imports the full realm, so we need to also change the client values in the realm itself.

michaelstingl commented:
Thanks for the feedback\! I've updated the `opencloud-realm.dist.json` file to synchronize all client IDs, redirect URIs, post-logout URIs, and other settings with the values in the individual client JSON files.

Since the realm configuration file is quite complex, please let me know if I missed anything else that needs to be aligned. This should ensure that when the realm is imported during deployment, all client configurations match exactly.

michaelstingl commented:
I noticed one more issue while testing and fixed it in another commit: 

The web client configuration also needed harmonization. I updated it to:
- Add the proper client name ("OpenCloud Web App")
- Use {{OC_URL}} template variables instead of hardcoded URLs
- Update redirect URIs to match exact paths instead of wildcard
- Update backchannel logout URL to use the template

These changes should complete the harmonization of all clients in the realm configuration.

micbar commented on deployments/examples/opencloud_full/config/keycloak/docker-entrypoint-override.sh:
```diff
@@ -2,7 +2,9 @@
 # replace openCloud domain in keycloak realm import
 mkdir /opt/keycloak/data/import
-sed -e "s/cloud.opencloud.test/${OC_DOMAIN}/g" /opt/keycloak/data/import-dist/opencloud-realm.json > /opt/keycloak/data/import/opencloud-realm.json
+sed -e "s/cloud.opencloud.test/${OC_DOMAIN}/g" \
```
What is the gain of that change?
Before your change, it was already a valid config, because `cloud.opencloud.test` is used on all testing and developer machines.
 ${OC_DOMAIN} is not always populated if you do not run the entrypoint script.

So i would like to keep it like it was.

micbar reviewed (changes requested):
Working fine. I am not so happy with the new OC_URL placeholder.

michaelstingl commented:
Thanks for the feedback about the template variables! I've reverted those changes to match the upstream conventions.

I initially added the {{OC_URL}} variables because I noticed they're used in the built-in IdP configuration (in services/idp/pkg/config/defaults/defaultconfig.go) and wanted to make the Keycloak configuration consistent with that pattern. However, I understand now that the Keycloak configuration should follow its own established convention with direct URLs and domain substitution.

I've:
1. Changed web client URLs back to direct format (https://cloud.opencloud.test)
2. Reverted docker-entrypoint-override.sh to only handle direct domain replacement

The important client ID harmonization and redirect URI updates remain in place. Let me know if there's anything else that should be adjusted!

micbar reviewed (approved):
Thank you! 🙏

suse-coder commented:
There are still old references like e4rAsNUSIUs0lF4nbv9FmCeUkTlV9GdgTLDH1b5uie7syb90SzEVrbN7HIpmWJeD in the new [/keycloak/opencloud-realm.dist.json](https://github.com/opencloud-eu/opencloud/pull/589/files#diff-30f37313608b98bed4d791621592bbed6bdfa6ced3455d48112ae0bbbed30b09)

michaelstingl commented:
Thanks @suse-coder for catching those remaining old client IDs! I've fixed all the references in the role mappings section of the configuration, replacing the long random IDs with the simplified client names.

This should complete the harmonization of client IDs throughout the entire realm configuration.
